### ISSUE TITLE
CSP

### ISSUE THREAD

**Issue by Dreamsorcerer** [mdn/web-components-examples#28]
I find the examples interesting, however I can't see how you would make these work with a reasonable CSP.

I rather like the idea of having CSS contained within a shadow DOM, however all your examples appear to use inline styles. Is there a way to combine this with a CSP in a reasonable manner?

Using external CSS files for each component could easily result in dozens of network requests, so that doesn't feel like a better solution. I'm also struggling to think of anyway to use hashes without some complex tools needing to be created (and I'm trying to avoid needing a build step). Any ideas on this?

**Comment by Leokuma** [comment]
The examples don't use inline styles. They create a `<style>` node inside the shadow DOM.

**Comment by Dreamsorcerer** [comment]
Which requires `'unsafe-inline'` to be added to the CSP in order to run them...

**Comment by Leokuma** [comment]
I understand your concern now. I don't really have an answer.

I just want to add something. An external CSS file for each component will be requested only once - no matter how many times your component is used in your website - because the browser will cache that CSS file.

Still, each _different_ component would require one network request, slowing down the first loading of your website.

**Comment by Dreamsorcerer** [comment]
Indeed, in my mind a component is the JS class, so that is what I intended to mean, as opposed to the number of custom elements used in the document. I'm thinking about using a number of components (maybe 20+) that may only be used in the page once each (like widgets for a dashboard), so I still need a way to avoid this.

**Comment by Dreamsorcerer** [comment]
The problem is that the examples do not work with a reasonable CSP, and there is no documentation that might suggest how you could make them work. This is still a limitation of the examples here, nothing to do with my code.
